service/history/shard/context_impl.go

```diff
@@ -836,35 +836,44 @@ func (s *ContextImpl) DeleteWorkflowExecution(
 
 	// Do not get namespace cache within shard lock.
 	namespaceEntry, err := s.GetNamespaceRegistry().GetNamespaceByID(namespace.ID(key.NamespaceID))
+	deleteVisibilityRecord := true
 	if err != nil {
-		return err
+		if _, isNotFound := err.(*serviceerror.NotFound); isNotFound {
+			// If namespace is not found, skip visibility record delete but proceed with other deletions.
+			// This case might happen during namespace deletion.
+			deleteVisibilityRecord = false
+		} else {
+			return err
+		}
 	}
 
 	s.wLock()
 	defer s.wUnlock()
 
 	// Step 1. Delete visibility.
-	addTasksRequest := &persistence.AddHistoryTasksRequest{
-		ShardID:     s.shardID,
-		NamespaceID: key.NamespaceID,
-		WorkflowID:  key.WorkflowID,
-		RunID:       key.RunID,
-
-		Tasks: map[tasks.Category][]tasks.Task{
-			tasks.CategoryVisibility: {
-				&tasks.DeleteExecutionVisibilityTask{
-					// TaskID is set by addTasksLocked
-					WorkflowKey:         key,
-					VisibilityTimestamp: s.timeSource.Now(),
-					Version:             newTaskVersion,
-					CloseTime:           closeTime,
+	if deleteVisibilityRecord {
```

Comment:
What if the visibility record is not deleted? Will it stay forever? Why not just proceed with delete visibility task, and just blindly issue request to delete visibility record from either db or ES.

Reply:
1. Visibility task creation requires namespace (mostly to get active cluster name and logging) and I don't want to change this logic.
2. This is not practically possible. If visibility record is not deleted, then namespace delete workflow "sees" this execution and will never delete namespace. Probably there is a very narrow edge case when visibility record is created after namespace is deleted but I can't even model this case.